crates/cubecl-ir/src/item.rs

```diff
@@ -301,37 +278,194 @@ impl Elem {
             },
-            Elem::UInt(kind) | Elem::AtomicUInt(kind) => match kind {
+            ElemType::UInt(kind) => match kind {
                 UIntKind::U8 => ConstantScalarValue::UInt(u8::MIN.into(), UIntKind::U8),
                 UIntKind::U16 => ConstantScalarValue::UInt(u16::MIN.into(), UIntKind::U16),
                 UIntKind::U32 => ConstantScalarValue::UInt(u32::MIN.into(), UIntKind::U32),
                 UIntKind::U64 => ConstantScalarValue::UInt(u64::MIN, UIntKind::U64),
             },
-            Elem::Bool => ConstantScalarValue::Bool(false),
+            ElemType::Bool => ConstantScalarValue::Bool(false),
         };
 
-        Variable::new(VariableKind::ConstantScalar(value), Item::new(*self))
+        Variable::new(VariableKind::ConstantScalar(value), Type::scalar(*self))
+    }
+}
+
+impl StorageType {
+    pub fn elem_type(&self) -> ElemType {
+        match self {
+            StorageType::Scalar(ty) | StorageType::Packed(ty, _) | StorageType::Atomic(ty) => *ty,
+            StorageType::Semantic(_) => unimplemented!("Semantic type has no elem type"),
+        }
+    }
+
+    pub fn packing_factor(&self) -> u32 {
+        match self {
+            StorageType::Packed(_, factor) => *factor,
+            _ => 1,
+        }
+    }
+
+    pub fn is_atomic(&self) -> bool {
+        matches!(self, StorageType::Atomic(_))
+    }
+
+    pub fn size(&self) -> usize {
+        self.size_bits().div_ceil(8)
+    }
+
+    pub fn size_bits(&self) -> usize {
+        match self {
+            StorageType::Packed(ty, factor) => ty.size_bits() * *factor as usize,
+            StorageType::Scalar(ty) | StorageType::Atomic(ty) => ty.size_bits(),
+            StorageType::Semantic(_) => 0,
+        }
+    }
+
+    /// Ensure that the variable provided, when a constant, is the same type as elem.
+    pub fn from_constant(&self, constant: Variable) -> Variable {
+        self.elem_type().from_constant(constant)
+    }
+
+    pub fn is_int(&self) -> bool {
+        self.elem_type().is_int()
+    }
+
+    pub fn is_signed_int(&self) -> bool {
+        self.elem_type().is_signed_int()
+    }
+
+    pub fn is_unsigned_int(&self) -> bool {
+        self.elem_type().is_unsigned_int()
+    }
+
+    pub fn is_float(&self) -> bool {
+        self.elem_type().is_float()
+    }
+}
+
+impl From<ElemType> for Type {
+    fn from(val: ElemType) -> Self {
+        Type::scalar(val)
+    }
+}
+
+impl From<ElemType> for StorageType {
+    fn from(val: ElemType) -> Self {
+        StorageType::Scalar(val)
+    }
+}
+
+impl From<SemanticType> for Type {
+    fn from(val: SemanticType) -> Self {
+        Type::new(StorageType::Semantic(val))
+    }
+}
+
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Debug, Clone, Copy, TypeHash, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct Type {
+    pub storage: StorageType,
+    pub line_size: LineSize,
+}
+
+pub type LineSize = Option<NonZero<u8>>;
```

Comment:
It should be a single `u8` and 0 is the same as not defined. `Some(0)` would be wrong right now, that's why we have `NonZero` but if `0 == None` we don't have that problem. The Option and NonZero is so unconvenient.